### PR TITLE
sys/shell: Exit the shell on ctrl-D

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

**Note** the control-C part of this PR was split to #11004 .

Right now the only way to exit the shell is if stdin is closed. This works on native, but on an embedded platform stdin is the uart and thus is never closed.

This patch causes the shell loop to exit on EOT (ASCII 0x04 / ctrl-D a.k.a. "End-of-Transmission".)

### Testing procedure

Run the test in `tests/shell` with `native` _and_ with a board.

```
$ cd tests/shell
$ BOARD=samr21-xpro make all
$ BOARD=samr21-xpro make flash
$ BOARD=samr21-xpro make test
```

To test it by hand with a board, use a terminal program that passes ctrl-D to the tty. In native, use Ctrl-V ctrl-D to insert a literal EOT.

### Related PRs

Depends on #11003 , #11004.

### CODE REVIEW

**Comment by kaspar030** [comment]
> on an embedded platform

use case?

**Comment by jcarrano** [comment]
> use case?

Examples, test scripts. In my case I was trying to integrate more tests in to `tests/shell` and needed to exit the shell.

Right now the shell is an infinite loop (except in native). I believe there should be a way to get out.

**Comment by kaspar030** [comment]
> needed to exit the shell.

Why?

**Comment by miri64** [comment]
I don't know about this. I sometimes use the fact in my testing, that the node is still accessible with `make term` after I `Strg+C` out of a previous `make term`.

> * Fix the use of python dictionaries: Python dictionaries are not guaranteed to be ordered until version
>   3.7. In 3.6 they are ordered too, but that is an implementation detail. riotdocker seems to be using 3.5. The commands in this test were stored in a dict.

There are also [OrderedDict](https://docs.python.org/2/library/collections.html#collections.OrderedDict).

**Comment by jcarrano** [comment]
> I don't know about this. I sometimes use the fact in my testing, that the node is still accessible with make term after I Strg+C out of a previous make term.

If you have a recent version of python everything should be OK. In 3.7 it is even documented that they preserver insertion order.

> There are also OrderedDict.

Yes, but the test here was only using a dict to store pairs and not doing indexing of any kind. In that case a list of 2-tuples is the best choice.

**Comment by miri64** [comment]
> > I don't know about this. I sometimes use the fact in my testing, that the node is still accessible with make term after I Strg+C out of a previous make term.
> 
> If you have a recent version of python everything should be OK. In 3.7 it is even documented that they preserver insertion order.

I don't see how this helps me with the shell being closed and me not being able to access the node anymore...

**Comment by miri64** [comment]
(unless I reset)

**Comment by jcarrano** [comment]
> I don't see how this helps me with the shell being closed and me not being able to access the node anymore...

Sorry, I messed up the comments.
This should not break your test. If you are using pyterm, Ctrl-C gets caught by pyterm, and causes pyterm to exit. In fact, it seems you cannot send through the Ctrl-C/D using pyterm.

**Comment by miri64** [comment]
> This should not break your test. If you are using pyterm, Ctrl-C gets caught by pyterm, and causes pyterm to exit. In fact, it seems you cannot send through the Ctrl-C/D using pyterm.

This still leaves the node in an "unexpected" state. If I reconnect with a terminal I expect the node still to do  whatever it was doing before (unless I toggle the reset pin), not being able to interact with it anymore, because the shell isn't there anymore.

You still don't answered the question why the shell needs to be exited.

**Comment by kaspar030** [comment]
> Lastly: _why the shell needs not to exit?_ I honestly don't understand why we are bikeshedding a <1 line change.

This is not bikeshedding. You stated "In my case I was trying to integrate more tests in to tests/shell and needed to exit the shell." and I asked to clarify why you need to exit the shell.

Even a one line change needs to be justified, especially in something that was in the same shape for almost a decade.

I'm not opposed to the change, just wondering why on earth you need to exit the shell in order to add more tests, and if there's not a simpler solution.

**Comment by benemorius** [comment]
I use an 'exit' command to close the shell. ^D is nice but I typed it too often without thinking and then my shell was gone. But it's true that wrapping it in a while loop solves this for applications where you don't want the shell to ever go away (for instance I think everything currently in examples would qualify). This would mimic how getty works on linux. You can always exit the shell, but by default it is configured to respawn immediately. 

I've needed to close the shell too. You have to if you want to use the uart for something else. I telnet in over the wireless interface and get a shell there, then run a command to kill the other shell running on the uart and redirect the uart through the telnet terminal. In this way I use a riot board as a wireless ftdi cable. Afterwards I respawn the shell on the uart.

This is one use case for closing the shell but certainly there are others.

File: sys/shell/shell.c [hunk 1]
```diff
@@ -227,7 +227,7 @@ static int readline(char *buf, size_t size)
         }
 
         int c = getchar();
-        if (c < 0) {
+        if (c < 0 || c == '\x03' || c == '\x04') {
```
Comment (benemorius) [inline sys/shell/shell.c]: I don't think it's right to exit on ^C. We should cancel the current line and print a new prompt. This is consistent with how bash and friends behave. Like this: benemorius@a946c44c7650d2ed3fea55d429485d7f8bdc30bf

Comment (jcarrano) [reply sys/shell/shell.c]: Thanks for the suggestion. I have implemented it, I think it is more useful.

**Comment by jcarrano** [comment]
My interpretation of the comments is that what's contested here is exiting with control-D. Cancelling with control-C should not be as controversial, right?

I will split this PR.

**Comment by jcarrano** [comment]
There's a couple of uses for this functionality which I had not thought of before:

- Simple user/pass login: One could put the shell in a while(1) loop, preceded by a login prompt function. This would only drop you to the shell if you had the correct user/password. After using the shell you go back to the login prompt by typing ctrl-d.
- What happens if the file object used by stdin is closed while there is a thread waiting on getchar? The shell gets an EOF and exits. This can happen if stdio is implemented in some way other than the UART (e.g. bluetooth). There we can see that the shell exiting is not such an exceptional circumstance.

**Comment by benemorius** [comment]
After re-reading my own comments I should clarify that I'm strongly in favor of exiting on ^D. We just need to wrap existing consumers of the shell with a while loop to avoid functional changes in existing code which doesn't wish to have a shell capable of exiting.

Does wrapping the shell in a while loop solve everyone's concerns with exiting on ^D?

**Comment by benemorius** [comment]
Does wrapping the shell in a while loop solve everyone's concerns with exiting on ^D?

**Comment by miri64** [comment]
> Does wrapping the shell in a while loop solve everyone's concerns with exiting on ^D?

So this is how `shell_run()` currently looks like:

https://github.com/RIOT-OS/RIOT/blob/711854562b77e2c9165942f9dc1a3ecef32dc8d0/sys/shell/shell.c#L293-L310

Currenlty, ^D is just ignored (`handle_input_line()` sees `*pos <= ' '` so nothing happens). The functionality introduced here causes the function to go into the if condition and break that loop. Adding another while loop around the existing while loop however would have the same behavior, just with another while loop around: Hitting ^D would finish the inner while loop, the outer while loop would then cause another iteration of the inner while loop. So I don't really get the point of having this then except for the `handle_input_line()` not being entered.

**Comment by benemorius** [comment]
Ok so when I say to wrap existing consumers of the shell with a while loop I don't mean adding a loop to `shell.c::shell_run()` but rather to `examples/default/main.c::main()` for example.

The idea is that we can make the shell optionally exitable but then modify all the existing callers of `shell_run()` to restore the current-master non-exitable behavior for whichever cases we don't actually want the shell to ever exit.

**Comment by miri64** [comment]
Ok, got it, but to quote yourself:

> We just need to wrap existing consumers of the shell with a while loop to avoid functional changes in existing code which doesn't wish to have a shell capable of exiting.

would mean to avoid functional changes we need to touch _every_ call of `shell_run` then... Given that we use that a lot, this seems somewhat of a big effort to me. As a compromise: how about introducing `shell_run_once()` to the shell's with the current functionality of `shell_run()` (+ the one introduced in this PR) and moving `shell_run()` to a function that calls `shell_run_once()` in an infinite loop?

**Comment by benemorius** [comment]
> Ok, got it, but to quote yourself:
> 
> > We just need to wrap existing consumers of the shell with a while loop to avoid functional changes in existing code which doesn't wish to have a shell capable of exiting.
> 
> would mean to avoid functional changes we need to touch _every_ call of `shell_run`

Yes, correct.

I mean grep says it's not that many changes and all of them are simply adding a while(1) loop which should be trivial to review. We *do* need an exitable shell. End users can make an exitable shell non-exitable but they cannot make a non-exitable shell exitable.

> As a compromise: how about introducing `shell_run_once()` to the shell's with the current functionality of `shell_run()` (+ the one introduced in this PR) and moving `shell_run()` to a function that calls `shell_run_once()` in an infinite loop?

At a glance, I have no objection to this.

**Comment by miri64** [comment]
> I mean grep says it's not that many changes and all of them are simply adding a while(1) loop which should be trivial to review.

Nevertheless, it's a moving target, as new tests or examples might introduce additional calls. In my experience those are the real trouble, as behavioral changes like this might take a while to settle for devs used to the current behavior.

**Comment by miri64** [comment]
> > As a compromise: how about introducing `shell_run_once()` to the shell's with the current functionality of `shell_run()` (+ the one introduced in this PR) and moving `shell_run()` to a function that calls `shell_run_once()` in an infinite loop?
> 
> At a glance, I have no objection to this.

See https://github.com/RIOT-OS/RIOT/pull/12272 for a demonstrator of what I mean.

**Comment by jcarrano** [comment]
One thing I must point out is that ctrl-D handling in native and on serial based targets is different (and that should be fixed some day).

Native uses the terminal in cooked mode, so it will never see the ctrl-d, instead it sees an EOF.

Embedded targets use what would be the equivalent of raw mode, thus seen ctrl-d, ctrl-c, etc as what they are (though depending on the terminal program you are using you may have to use Ctrl-V to prevent it from being interpreted by your *local* machine).

IMO this discrepancy should be fixed, and native should be setting up the terminal to behave as close to the UART as possible (raw mode, no buffering, no echo) using `tcsetattr` and friends.

All this being said, it may make sense to differentiate between the stream being closed and a ctrl-d (maybe do a `feof()` before returning from shell_run? or add an additional code to handle_input_line?) and change the return type of shell run to allow it to report this.

**Comment by jcarrano** [comment]
> ... shell_run_once() ...
> .. moving shell_run() to a function that calls shell_run_once() ...

you can do this, but why invent new "sentences" when you can express that same logic using the vocabulary already present in C (`while (stream not closed) { ... shell_run(...) .. }`)

**Comment by miri64** [comment]
Alternative approach provided in #13675.

**Comment by benpicco** [comment]
@miri64 this is unrelated.
On `native` the shell never receives the raw `EOT`. The signal is caught by the operating system and results in closing `stdin`, so the shell will exit.
This PR is about replicating this behavior on the target.

**Comment by miri64** [comment]
Ok

**Comment by fjmolinas** [comment]
Now that the examples uses `shell_run()` can we move forward with this one? @benemorius are you still interested?, @HendrikVE you have been modifying the shell a lot lately?

**Comment by HendrikVE** [comment]
@fjmolinas Yes, right now `ctrl-D` is not caught at all, leading to consuming it as a normal character. This is not good, it should at least caught and ignored. But now, as we have the infinite loop wrapper around the shell it might be less controversal to exit the shell with `ctrl-D`?

**Comment by HendrikVE** [comment]
With the new shell it would look like #14345

**Comment by miri64** [comment]
Then let's close this?

**Pull request closed without merging**